File: cpp/RMXKit/NotificationCenter.cpp
//
//  RMXNotificationCenter.cpp
//  RMXKit
//
//  Created by Max Bilbow on 28/07/2015.
//  Copyright © 2015 Rattle Media Ltd. All rights reserved.
//


#import <iostream>
//#import "EventListener.hpp"
//#import "LinkedList.hpp"
//#import "Dictionary.hpp"
#import "NotificationCenter.hpp"


using namespace rmx;

typedef NotificationCenter::EventArgs EventArgs;
typedef NotificationCenter::EventType EventType;
typedef NotificationCenter::EventStatus EventStatus;




using namespace std;

NotificationCenter::NotificationCenter() {
    listeners = new LinkedList<EventListener>();
    events = new Dictionary<EventType, EventStatus>();
}

static bool _init;
static NotificationCenter * _current;

NotificationCenter * NotificationCenter::current() {
    if (!_init) {
        _current = new NotificationCenter();
        _init = true;
    }
    return _current;
}
    
bool NotificationCenter::hasListener(EventListener * listener) {
     return current()->listeners->contains(listener);// Listeners.Contains (listener);
}

void NotificationCenter::reset(EventType theEvent) {
//    theEvent->status = Idle;
    current()->events->setValueForKey(theEvent, EVENT_STATUS_IDLE);
}
    
void NotificationCenter::addListener(EventListener * listener) {
    if (!current()->listeners->contains (listener)) {
        current()->listeners->append(listener);
//        cout << "\nAdded to Listeners: " << *listener << endl;
    } else {
        cout << "\n !! Not Adding => Listeners already contain: " << *listener << endl;
    }
}

EventListener * NotificationCenter::removeListener(EventListener * listener) {
    return current()->listeners->removeValue(listener);
}

EventStatus NotificationCenter::statusOf(EventType theEvent) {
    try {
        return *current()->events->getValueForKey(theEvent);
    } catch (exception e) {
        cout << e.what() << ": setting new event as Idle" << endl;
        current()->events->setValueForKey(theEvent, EVENT_STATUS_IDLE);
        return EVENT_STATUS_IDLE;
    }
}

bool NotificationCenter::isIdle(EventType theEvent) {
    return statusOf(theEvent) == EVENT_STATUS_IDLE;
}

 bool NotificationCenter::isActive(EventType theEvent) {
    return statusOf(theEvent) == EVENT_STATUS_ACTIVE;
}


void NotificationCenter::eventDidOccur(EventType theEvent, EventArgs o) {
    eventWillStart(theEvent,o);
    eventDidEnd(theEvent,o);
}


bool NotificationCenter::isComplete(EventType theEvent) {
    return statusOf(theEvent) == EVENT_STATUS_COMPLETE;
}

bool NotificationCenter::didFail(EventType theEvent) {
    return statusOf(theEvent) == EVENT_STATUS_FAILURE;
}


void NotificationCenter::eventWillStart(EventType theEvent, EventArgs o) {
    LinkedList<EventListener>::Node * node = current()->listeners->firstNode();
    if (node->value == nullptr)
        return;
    else {
        while (node != nullptr) {
            node->value->OnEventDidStart(theEvent,o);
            node = node->next;
        }
    }
}

void NotificationCenter::eventDidEnd(EventType theEvent, EventArgs o) {
    LinkedList<EventListener>::Node * node = current()->listeners->firstNode();
    if (node->value == nullptr)
        return;
    else {
        while (node != nullptr) {
            node->value->OnEventDidEnd(theEvent, o);
            node = node->next;
        }
    }
}

 void NotificationCenter::notifyListeners(string message) {
     LinkedList<EventListener>::Node * node = current()->listeners->firstNode();
     if (node->value == nullptr)
         return;
     else {
         while (node != nullptr) {
             node->value->SendMessage(message);
             node = node->next;
         }
     }
}

void NotificationCenter::Test() {
    (new EventListener())->SendMessage("Yo!");
    NotificationCenter::eventWillStart("Hello, world!");
    NotificationCenter::eventDidEnd("Googbye, world!");
}
    

File: Java/src/rmx/NotificationCenter.java
package rmx;

import java.util.Dictionary;

public class NotificationCenter {
	private static Dictionary<String,EventListener> listeners;// = new Dictionary<String,EventListener> ();

	private static Dictionary<String,IEvent> events;// = new Dictionary<IEvent,EventStatus>();






	public static boolean HasListener(EventListener listener) {
		return listeners.get(listener.name()) != null;
	}
	
	public static void Reset(IEvent theEvent) {
		if (events.get(theEvent.type()) != null)
			events.get(theEvent.type()).setStatus(EventStatus.Idle);
		else {
			theEvent.setStatus(EventStatus.Idle);
			events.put(theEvent.type(),theEvent);
		}
	}

	public static void AddListener(EventListener listener) {
		listeners.put(listener.Name) = listener;
		if (Bugger.WillLog (RMXTests.EventCenter, listener.GetType () + " was added to Listeners ("+ Listeners.Count + ")"))
			Debug.Log (Bugger.Last);
//		if (Bugger.WillLog(Testing.EventCenter, "Listeners: " + Listeners.Count))
//			Debug.Log (Bugger.Last);

	}

	public static void RemoveListener(EventListener listener) {
		if (Listeners.ContainsValue (listener))
		if (!Listeners.Remove (listener.name))
			throw new System.Exception (listener.name + " exists but could not be removed from Listeners");
	}

	public static EventStatus StatusOf(IEvent theEvent) {
		return Events.ContainsKey(theEvent) ? Events [theEvent] : EventStatus.Idle;
	}

	public static bool IsIdle(IEvent theEvent) {
		return StatusOf (theEvent) == EventStatus.Idle;
	}

	public static bool IsActive(IEvent theEvent) {
		return StatusOf (theEvent) == EventStatus.Active;
	}

	public static void EventDidOccur(IEvent e) {
		EventDidOccur (e, null);
	}

	public static void EventDidOccur(IEvent theEvent, object o) {
		var listeners = Listeners;
		Events [theEvent] = o is EventStatus ? (EventStatus) o : EventStatus.Completed;
		foreach (KeyValuePair<string, EventListener> listener in listeners) {
			listener.Value.OnEvent(theEvent,o);
		}
	}


	public static bool WasCompleted(IEvent theEvent) {
		return StatusOf (theEvent) == EventStatus.Completed;
	}

	public static void EventWillStart(IEvent theEvent) {
		EventWillStart (theEvent, null);
	}

	public static void EventWillStart(IEvent theEvent, object o) {
		if (!IsActive (theEvent)) {
			Events [theEvent] = o is EventStatus ? (EventStatus) o : EventStatus.Active;
			foreach (KeyValuePair<string, EventListener> listener in Listeners) {
				listener.Value.OnEventDidStart (theEvent, o);
			}
		}
	}
	public static void EventDidEnd(IEvent theEvent) {
		EventDidEnd (theEvent, null);
	}
	public static void EventDidEnd(IEvent theEvent, object o) {
		var listeners = Listeners;
		Events [theEvent] = Events [theEvent] = o is EventStatus ? (EventStatus) o : EventStatus.Completed;
		foreach (KeyValuePair<string, EventListener> listener in listeners) {
			listener.Value.OnEventDidEnd(theEvent,o);
		}
	}

	public static void NotifyListeners(string message) {
		foreach (KeyValuePair<string, EventListener> listener in Listeners) {
			listener.Value.SendMessage (message, SendMessageOptions.DontRequireReceiver);
		}
	}

}
